### PR TITLE
Reading annotation files by pattern

### PR DESCRIPTION
**Pull request by Ilya-Krylov**

### CODE REVIEW

**Comment by AlexanderDokuchaev** [comment]
run tests

File: mmdet/datasets/builder.py [hunk 1]
```diff
@@ -35,6 +55,20 @@ def build_dataset(cfg, default_args=None):
             build_dataset(cfg['dataset'], default_args), cfg['times'])
     elif isinstance(cfg['ann_file'], (list, tuple)):
         dataset = _concat_dataset(cfg, default_args)
+    elif '*' in cfg['ann_file']:
```
Comment (druzhkov-paul) [inline mmdet/datasets/builder.py]: Maybe it'd be better to do globbing first and then process as usual?

**Review by druzhkov-paul — CHANGES_REQUESTED** [review]

File: mmdet/datasets/builder.py (hunk 1 above)
Comment (Ilya-Krylov) [reply mmdet/datasets/builder.py]: changed to glob, but * is required to avoid endless recursion of `build_dataset`

Comment (druzhkov-paul) [reply mmdet/datasets/builder.py]: I thought about something like this:
```python
else:
        cfg['ann_file'] = glob.glob(cfg['ann_file'], recursive=True)
        assert len(cfg['ann_file']) > 0
        if len(cfg['ann_file']) == 1:
            cfg['ann_file'] = cfg['ann_file'][0]
            dataset = build_from_cfg(cfg, DATASETS, default_args)
        else:
            dataset = _concat_dataset(cfg, default_args)
```
What do you think?

File: mmdet/datasets/builder.py [hunk 2]
```diff
@@ -11,6 +13,24 @@ def _concat_dataset(cfg, default_args=None):
...
+            # dataset_root
+            # ├── annotations
+            # │   ├── instances_train.json
+            # │   ├── ...
+            # ├── images
+            #     ├── image_name1
+            #     ├── image_name2
+            #     ├── ...
+            # and file_name inside instances_train.json is relative to dataset root
+            img_prefixes = [os.path.join(os.path.dirname(ann_file), '..') for ann_file in ann_files]
```
Comment (druzhkov-paul) [inline mmdet/datasets/builder.py]: Shouldn't the path contain `images` as well?

Comment (Ilya-Krylov) [reply mmdet/datasets/builder.py]: this is a matter of some agreement. Added

File: mmdet/datasets/builder.py (hunk 1 above)
Comment (Ilya-Krylov) [reply mmdet/datasets/builder.py]: fixed

File: mmdet/datasets/builder.py [hunk 3]
```diff
@@ -36,6 +58,15 @@ def build_dataset(cfg, default_args=None):
     elif isinstance(cfg['ann_file'], (list, tuple)):
         dataset = _concat_dataset(cfg, default_args)
     else:
-        dataset = build_from_cfg(cfg, DATASETS, default_args)
+        matches = glob.glob(cfg['ann_file'], recursive=True)
+        if not matches:
+            raise RuntimeError(f'Failed to fined annotation files that match pattern: '
```
Comment (druzhkov-paul) [inline mmdet/datasets/builder.py]: `find`

File: mmdet/datasets/builder.py [hunk 4]
```diff
@@ -11,6 +14,25 @@ def _concat_dataset(cfg, default_args=None):
...
+            # ├── annotations
+            # │   ├── instances_train.json
+            # │   ├── ...
+            # ├── images
+            #     ├── image_name1
+            #     ├── image_name2
+            #     ├── ...
+            # and file_name inside instances_train.json is relative to <dataset_root>/images
+            img_prefixes = \
+                [os.path.join(os.path.dirname(ann_file), '../images') for ann_file in ann_files]
```
Comment (druzhkov-paul) [inline mmdet/datasets/builder.py]:
```python
os.path.join(os.path.dirname(ann_file), '..', 'images')
```

**Review by druzhkov-paul — APPROVED** [review]

File: mmdet/datasets/builder.py [hunk 5]
```diff
@@ -36,6 +58,15 @@ def build_dataset(cfg, default_args=None):
     elif isinstance(cfg['ann_file'], (list, tuple)):
         dataset = _concat_dataset(cfg, default_args)
     else:
-        dataset = build_from_cfg(cfg, DATASETS, default_args)
+        matches = glob.glob(cfg['ann_file'], recursive=True)
+        if not matches:
+            raise RuntimeError(f'Failed to fine annotation files that match pattern: '
```
Comment (druzhkov-paul) [inline mmdet/datasets/builder.py]: `find` :smile:

**Pull request merged**